File: day03/main.scala
import scala.collection.mutable.ArrayBuffer

case class Point(val x: Int, val y: Int) {
  def toTuple: (Int, Int) =
    (x, y)
}

object Day03 {
  def findClosestDistance(wire1: ArrayBuffer[Point], wire2: ArrayBuffer[Point], initialPoint: Point): Int = {
    var lowestDistance = 0
    val intersections = wire1.intersect(wire2)

    for (point <- intersections) {
      val auxDistance = distance(initialPoint, point)

      if (lowestDistance == 0 || auxDistance < lowestDistance) {
        lowestDistance = auxDistance
      }
    }

    lowestDistance
  }

  def findMinSteps(wire1: ArrayBuffer[Point], wire2: ArrayBuffer[Point], initialPoint: Point): Int = {
    var minSteps = 0
    val intersections = wire1.intersect(wire2)

    for (point <- intersections) {
      val steps = wire1.indexOf(point) + wire2.indexOf(point) + 2

      if (minSteps == 0 || steps < minSteps) {
        minSteps = steps
      }
    }

    minSteps
  }

  def distance(p1: Point, p2: Point): Int = {
    math.abs(p1.x - p2.x) + math.abs(p1.y - p2.y)
  }

  def navigate(positions: ArrayBuffer[Point], directions: Seq[String], current: Int, currentPosition: Point): Unit = {
    if (current >= directions.length)
      return

    val (direction: String, steps: Int) = parseDirection(directions(current))

    val newPosition = direction match {
      case "U" => goUp(positions, currentPosition, steps)
      case "D" => goDown(positions, currentPosition, steps)
      case "L" => goLeft(positions, currentPosition, steps)
      case "R" => goRight(positions, currentPosition, steps)
    }

    navigate(positions, directions, current + 1, newPosition)
  }

  def goUp(positions: ArrayBuffer[Point], currentPosition: Point, steps: Int): Point = {
    val (x, y) = currentPosition.toTuple
    for (i <- y + 1 to y + steps) {
      positions += new Point(x, i)
    }

    new Point(x, y + steps)
  }

  def goLeft(positions: ArrayBuffer[Point], currentPosition: Point, steps: Int): Point = {
    val (x, y) = currentPosition.toTuple
    for (i <- x - 1 to x - steps by -1) {
      positions += new Point(i, y)
    }

    new Point(x - steps, y)
  }

  def goDown(positions: ArrayBuffer[Point], currentPosition: Point, steps: Int): Point = {
    val (x, y) = currentPosition.toTuple
    for (i <- y - 1 to y - steps by -1) {
      positions += new Point(x, i)

    }

    new Point(x, y - steps)
  }

  def goRight(positions: ArrayBuffer[Point], currentPosition: Point, steps: Int): Point = {
    val (x, y) = currentPosition.toTuple
    for (i <- x + 1 to x + steps) {
      positions += new Point(i, y)
    }

    new Point(x + steps, y)
  }

  def parseDirection(direction: String): (String, Int) = {
    val steps = direction.slice(1, direction.length).toInt

    (direction(0).toString, steps)
  }

  def parseWireDirections(wire: String): Seq[String] = {
    wire.split(",").toSeq.map(_.trim).filter(_ != "")
  }

  def main(args: Array[String]): Unit = {
    val lines = scala.io.Source.fromFile("input").mkString
    val wires = lines.split("\n")
      .toSeq
      .map(_.trim)
      .filter(_ != "")
    // val wires = Array(
    //   "R8,U5,L5,D3",
    //   "U7,R6,D4,L4"
    // )
    // val wires = Array(
    //   "R75,D30,R83,U83,L12,D49,R71,U7,L72",
    //   "U62,R66,U55,R34,D71,R55,D58,R83"
    // )
    // val wires = Array(
    //   "R98,U47,R26,D63,R33,U87,L62,D20,R33,U53,R51",
    //   "U98,R91,D20,R16,D67,R40,U7,R15,U6,R7"
    // )
    val directions = wires.map(parseWireDirections)

    val initial = new Point(0, 0)
    val wire1: ArrayBuffer[Point] = ArrayBuffer[Point]()
    val wire2: ArrayBuffer[Point] = ArrayBuffer[Point]()

    navigate(wire1, directions(0), 0, initial)
    navigate(wire2, directions(1), 0, initial)

    val shortestDistance = findClosestDistance(wire1, wire2, initial)
    val minSteps = findMinSteps(wire1, wire2, initial)

    println(shortestDistance)
    println(minSteps)
  }
}

File: day03/main.rs
use std::collections::HashSet;

struct Instruction {
    direction: String,
    steps: i32,
}

#[derive(Debug, Clone, PartialEq, Eq, Hash)]
struct Point {
    x: i32,
    y: i32,
}

fn navigate(directions: &Vec<Instruction>) -> Vec<Point> {
    let mut current = Point { x: 0, y: 0 };
    let mut positions = Vec::new();

    for dir in directions {
        match dir.direction.as_ref() {
            "U" => {
                for i in (current.y + 1)..(current.y + dir.steps + 1) {
                    positions.push(Point { x: current.x, y: i });
                }
                current.y = positions.last().unwrap().y
            }
            "D" => {
                for i in ((current.y - dir.steps)..(current.y)).rev() {
                    positions.push(Point { x: current.x, y: i });
                }
                current.y = positions.last().unwrap().y
            }
            "R" => {
                for i in (current.x + 1)..(current.x + dir.steps + 1) {
                    positions.push(Point { x: i, y: current.y });
                }
                current.x = positions.last().unwrap().x
            }
            "L" => {
                for i in ((current.x - dir.steps)..(current.x)).rev() {
                    positions.push(Point { x: i, y: current.y });
                }
                current.x = positions.last().unwrap().x
            }
            _ => ()
        }
    }

    positions
}

fn intersection(wire1: &Vec<Point>, wire2: &Vec<Point>) -> Vec<Point> {
    let mut intersections: Vec<Point> = Vec::new();
    let mut position_set = HashSet::new();

    for position in wire1 {
        position_set.insert(position.clone());
    }

    for position in wire2 {
        if position_set.contains(&position) {
            intersections.push(position.clone());
        }
    }

    intersections
}

fn minimum_distance(wire1: &Vec<Point>, wire2: &Vec<Point>) -> i32 {
    let mut lowest = 0;
    let initial = Point { x: 0, y: 0 };
    let intersections = intersection(&wire1, &wire2);

    for point in intersections {
        let distance = manhatan_distance(&initial, &point);

        if lowest == 0 || distance < lowest {
            lowest = distance
        }
    }

    lowest
}

fn manhatan_distance(p1: &Point, p2: &Point) -> i32 {
    (p1.x - p2.x).abs() + (p1.y - p2.y).abs()
}

fn minimum_steps(wire1: &Vec<Point>, wire2: &Vec<Point>) -> i32 {
    let mut lowest: i32 = 0;
    let intersections = intersection(&wire1, &wire2);

    for point in intersections {
        let steps1 = wire1.iter().position(|w| w == &point).unwrap() + 1;
        let steps2 = wire2.iter().position(|w| w == &point).unwrap() + 1;
        let steps = (steps1 + steps2) as i32;

        if lowest == 0 || steps < lowest {
            lowest = steps
        }
    }

    lowest
}

fn parse(wire: String) -> Vec<Instruction> {
    wire.split(",")
        .map(|w| {
            let arr: Vec<_> = w.chars().collect();

            let direction: String = arr[0].to_string();
            let steps: String = arr[1..arr.len()].to_vec().iter().collect();

            Instruction { direction: direction, steps: steps.parse().unwrap() }
        })
        .collect()
}

fn main() {
    // let wire1 = "R75,D30,R83,U83,L12,D49,R71,U7,L72".to_string();
    // let wire2 = "U62,R66,U55,R34,D71,R55,D58,R83".to_string();
    // let wire1 = "R98,U47,R26,D63,R33,U87,L62,D20,R33,U53,R51".to_string();
    // let wire2 = "U98,R91,D20,R16,D67,R40,U7,R15,U6,R7".to_string();
    let input = std::fs::read_to_string("input")
        .expect("Unable to read file");

    let data: Vec<String> = input.trim()
        .split("\n")
        .map(|s| s.to_string())
        .collect();

    let wire1 = data[0].to_string();
    let wire2 = data[1].to_string();

    let wire1_instructions = parse(wire1);
    let wire2_instructions = parse(wire2);

    let wire1_positions = navigate(&wire1_instructions);
    let wire2_positions = navigate(&wire2_instructions);

    let answer1 = minimum_distance(&wire1_positions, &wire2_positions);
    let answer2 = minimum_steps(&wire1_positions, &wire2_positions);

    println!("{:?}", answer1);
    println!("{:?}", answer2);
}
